Build fails on Redhat 6.8 with GCC 4.4.7

## Steps to Reproduce
On a machine with GCC 4.4.7 as the default compiler (i.e. stock Centos6.9)
1. clone mono master
2. configure
3. make

### Current Behavior

Build fails.

### Expected Behavior

Build completes.

## On which platforms did you notice this

[ ] macOS
[ X ] Linux
[ ] Windows

**Version Used**:

```
$ git branch -v
* master dbfbffb559a Bump API snapshot submodule
$ gcc --version | head -1
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17)
$ uname -srv
Linux 2.6.32-696.3.2.el6.x86_64 #1 SMP Wed Jun 7 11:51:39 EDT 2017
```

## Stacktrace

```
trace.c: In function ‘mono_trace_enter_method’:
trace.c:146: error: #pragma GCC diagnostic not allowed inside functions
trace.c:148: error: #pragma GCC diagnostic not allowed inside functions
make[4]: *** [libmini_la-trace.lo] Error 1
make[4]: Leaving directory `/scratch/nlrvmp/mono/mono-git/mono/mini'
make[3]: *** [all] Error 2
make[3]: Leaving directory `/scratch/nlrvmp/mono/mono-git/mono/mini'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/scratch/nlrvmp/mono/mono-git/mono'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/scratch/nlrvmp/mono/mono-git'
make: *** [all] Error 2
```

Commenting out lines 146 and 148 gets me a successful build. I think this was introduced with #8308.

--- a/mono/mini/trace.c
+++ b/mono/mini/trace.c
@@ -143,10 +143,26 @@ mono_trace_enter_method (MonoMethod *method, char *ebp)
 	g_free (fname);
 
 	if (!ebp) {
+
+// https://github.com/mono/mono/issues/8572
+// Build fails on Redhat 6.8 with GCC 4.4.7.
+// Old gcc does not allow pragma diagnostic within function.
+#if __clang__ || __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
+#define GCC_PRAGMA_DIAGNOSTIC_WITHIN_FUNCTION 1
+#endif
+
+#if GCC_PRAGMA_DIAGNOSTIC_WITHIN_FUNCTION
 #pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunknown-warning-option"
 #pragma GCC diagnostic ignored "-Wframe-address"
+#endif
+
 		printf (") ip: %p\n", MONO_RETURN_ADDRESS_N (1));
+
+#if GCC_PRAGMA_DIAGNOSTIC_WITHIN_FUNCTION
 #pragma GCC diagnostic pop
+#endif
+
 		goto unlock;
 	}
 

